File: android/aspectly-bridge/src/main/kotlin/com/aspectly/bridge/BridgeHandler.kt
package com.aspectly.bridge

import com.google.gson.Gson
import com.google.gson.JsonElement

/** A raw handler: receives the request params as a [JsonElement] and returns an encodable result. */
typealias RawBridgeHandler = suspend (JsonElement) -> Any?

/**
 * Interface for type-safe bridge handlers. The Kotlin equivalent of .NET's /
 * Swift's `BridgeHandler`.
 */
interface BridgeHandler<P, R> {
    /** The method name that JavaScript will use to call this handler. */
    val methodName: String

    /** Handle the request with the deserialized parameters. */
    suspend fun handle(params: P): R
}

/** Shared Gson instance used for bridge (de)serialization. */
object BridgeJson {
    val gson: Gson = Gson()
}

File: swift/Sources/AspectlyBridge/BridgeHandler.swift
import Foundation

/// A handler closure registered with `BridgeHost`. Receives the raw request
/// params and returns an `Encodable` result (or `nil`).
public typealias RawBridgeHandler = (JSONValue) async throws -> Encodable?

/// Protocol for type-safe bridge handlers.
///
/// Implement this protocol to create handlers that can be registered with
/// `BridgeHost`. The Swift equivalent of .NET's `IBridgeHandler<TParams, TResult>`.
public protocol BridgeHandler {
    /// Parameters the handler accepts (decoded from the JS request).
    associatedtype Params: Decodable
    /// Result the handler returns (encoded back to JS).
    associatedtype Result: Encodable

    /// The method name that JavaScript will use to call this handler.
    var methodName: String { get }

    /// Handle the request with the given parameters.
    /// - Parameter params: Deserialized parameters from JavaScript.
    /// - Returns: The result to send back to JavaScript.
    func handle(_ params: Params) async throws -> Result
}
